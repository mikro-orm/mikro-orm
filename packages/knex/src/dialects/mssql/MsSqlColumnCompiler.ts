import { MonkeyPatchable } from '../../MonkeyPatchable';

export class MsSqlColumnCompiler extends MonkeyPatchable.MsSqlColumnCompiler {

  enu(this: any, allowed: unknown[]) {
    return `nvarchar(100) check (${this.formatter.wrap(this.args[0])} in ('${(allowed.join("', '"))}'))`;
  }

}
