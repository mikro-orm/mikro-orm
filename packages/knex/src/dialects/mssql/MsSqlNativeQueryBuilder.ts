import { LockMode, QueryFlag, RawQueryFragment, Utils } from '@mikro-orm/core';
import { NativeQueryBuilder } from '../../query/NativeQueryBuilder.js';
import { QueryType } from '../../query/enums.js';

/** @internal */
export class MsSqlNativeQueryBuilder extends NativeQueryBuilder {

  override compile(): { sql: string; params: unknown[] } {
    if (!this.type) {
      throw new Error('No query type provided');
    }

    this.parts.length = 0;
    this.params.length = 0;

    if (this.options.flags?.has(QueryFlag.IDENTITY_INSERT)) {
      this.parts.push(`set identity_insert ${this.getTableName()} on;`);
    }

    if (this.options.comment) {
      this.parts.push(...this.options.comment.map(comment => `/* ${comment} */`));
    }

    if (this.options.onConflict && !Utils.isEmpty(Utils.asArray(this.options.data)[0])) {
      this.compileUpsert();
    } else {
      switch (this.type) {
        case QueryType.SELECT:
        case QueryType.COUNT: this.compileSelect(); break;
        case QueryType.INSERT: this.compileInsert(); break;
        case QueryType.UPDATE: this.compileUpdate(); break;
        case QueryType.DELETE: this.compileDelete(); break;
        case QueryType.TRUNCATE: this.compileTruncate(); break;
      }

      if ([QueryType.INSERT, QueryType.UPDATE, QueryType.DELETE].includes(this.type)) {
        this.parts[this.parts.length - 1] += '; select @@rowcount;';
      }
    }

    if (this.options.flags?.has(QueryFlag.IDENTITY_INSERT)) {
      this.parts.push(`set identity_insert ${this.getTableName()} off;`);
    }

    return this.combineParts();
  }

  private compileUpsert() {
    const clause = this.options.onConflict!;
    const dataAsArray = Utils.asArray(this.options.data);
    const keys = Object.keys(dataAsArray[0]);
    const values = keys.map(() => '?');
    const parts = [];

    for (const data of dataAsArray) {
      for (const key of keys) {
        this.params.push(data![key]);
      }

      parts.push(`(${values.join(', ')})`);
    }

    this.parts.push(`merge into ${this.getTableName()}`);
    this.parts.push(`using (values ${parts.join(', ')}) as tsource(${keys.map(key => this.quote(key)).join(', ')})`);

    if (clause.fields instanceof RawQueryFragment) {
      this.parts.push(clause.fields.sql);
      this.params.push(...clause.fields.params);
    } else if (clause.fields.length > 0) {
      // const fields = clause.fields.map(field => this.quote(field));
      const conflictColumn = this.quote(clause.fields[0]);
      this.parts.push(`on ${this.getTableName()}.${conflictColumn} = tsource.${conflictColumn}`);
    }

    const sourceColumns = keys.map(field => `tsource.${this.quote(field)}`).join(', ');
    const destinationColumns = keys.map(field => this.quote(field)).join(', ');

    this.parts.push(`when not matched then insert (${destinationColumns}) values (${sourceColumns})`);

    if (!clause.ignore) {
      this.parts.push('when matched');

      if (clause.where) {
        this.parts.push(`and ${clause.where.sql}`);
        this.params.push(...clause.where.params);
      }

      this.parts.push('then update set');

      if (!clause.merge || Array.isArray(clause.merge)) {
        const parts = keys.map((column: any) => `${this.quote(column)} = tsource.${this.quote(column)}`);
        this.parts.push(parts.join(', '));
      } else if (typeof clause.merge === 'object') {
        const parts = Object.entries(clause.merge).map(([key, value]) => {
          this.params.push(value);
          return `${this.getTableName()}.${this.quote(key)} = ?`;
        });
        this.parts.push(parts.join(', '));
      }
    }

    this.addOutputClause('inserted');
    this.parts[this.parts.length - 1] += ';';
  }

  protected override compileSelect() {
    this.parts.push('select');

    if (this.options.limit != null && this.options.offset == null) {
      this.parts.push(`top (?)`);
      this.params.push(this.options.limit);
    }

    this.addHintComment();
    this.parts.push(`${this.getFields()} from ${this.getTableName()}`);
    this.addLockClause();

    if (this.options.joins) {
      for (const join of this.options.joins) {
        this.parts.push(join.sql);
        this.params.push(...join.params);
      }
    }

    if (this.options.where?.sql.trim()) {
      this.parts.push(`where ${this.options.where.sql}`);
      this.params.push(...this.options.where.params);
    }

    if (this.options.groupBy) {
      const fields = this.options.groupBy.map(field => this.quote(field));
      this.parts.push(`group by ${fields.join(', ')}`);
    }

    if (this.options.having) {
      this.parts.push(`having ${this.options.having.sql}`);
      this.params.push(...this.options.having.params);
    }

    if (this.options.orderBy) {
      this.parts.push(`order by ${this.options.orderBy}`);
    }

    if (this.options.offset != null) {
      /* v8 ignore next 3 */
      if (!this.options.orderBy) {
        throw new Error('Order by clause is required for pagination');
      }

      this.parts.push(`offset ? rows`);
      this.params.push(this.options.offset);

      if (this.options.limit != null) {
        this.parts.push(`fetch next ? rows only`);
        this.params.push(this.options.limit);
      }
    }
  }

  protected override addLockClause() {
    if (!this.options.lockMode || ![LockMode.PESSIMISTIC_READ, LockMode.PESSIMISTIC_WRITE].includes(this.options.lockMode)) {
      return;
    }

    const map = {
      [LockMode.PESSIMISTIC_READ]: 'with (holdlock)',
      [LockMode.PESSIMISTIC_WRITE]: 'with (updlock)',
    } as const;

    if (this.options.lockMode !== LockMode.OPTIMISTIC) {
      this.parts.push(map[this.options.lockMode as keyof typeof map]);
    }
  }

  protected override compileTruncate() {
    const tableName = this.getTableName();
    const sql = `delete from ${tableName}; declare @count int = case @@rowcount when 0 then 1 else 0 end; dbcc checkident ('${tableName.replace(/[[\]]/g, '')}', reseed, @count)`;
    this.parts.push(sql);
  }

}
