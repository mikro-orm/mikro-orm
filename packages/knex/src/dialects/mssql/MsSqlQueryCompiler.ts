/* istanbul ignore file */
import { MonkeyPatchable } from '../../MonkeyPatchable';
import { Utils } from '@mikro-orm/core';

// upsert support from https://github.com/knex/knex/pull/6050
export class MsSqlQueryCompiler extends MonkeyPatchable.MsSqlQueryCompiler {

  constructor(client: any, builder: any, formatter: any) {
    const onConflict = builder._single.onConflict;
    delete builder._single.onConflict;
    super(client, builder, formatter);
    (this as any).single.onConflict = onConflict;
  }

  // Compiles an "insert" query, allowing for multiple
  // inserts using a single query statement.
  insert(this: any) {
    if (this.single.onConflict) {
      return this._insertWithMerge();
    }

    return super.insert();
  }

  _mergeAnd(this: any) {
    const wheres = this.where();

    if (!wheres) {
      return '';
    }

    return `and ${wheres.slice(6)} `;
  }

  _mergeWhenMatched(this: any, columns: any, updates: any) {
    let columnsData: any = [];

    if (!updates || Array.isArray(updates)) {
      columnsData = columns
        .map((column: any) => `${this.formatter.columnize(column)}=tsource.${this.formatter.columnize(column)}`)
        .join(', ');
    }

    if (typeof updates === 'string') {
      columnsData = `${this.formatter.columnize(updates)}=tsource.${this.formatter.columnize(updates)}`;
    }

    if (!Array.isArray(updates) && typeof updates === 'object') {
      columnsData = Object.entries(updates)
        .map(([key, value]) => `${this.tableName}.${this.formatter.columnize(key)}=(${this._getParameters([value])})`);
    }

    const sql =  ` when matched ${this._mergeAnd()}then update set ${columnsData}`;

    return sql;
  }

  _mergeWhenNotMatched(this: any, columns: any) {
    const destinationColumns = this.formatter.columnize(columns);
    const sourceColumns = this.formatter.columnizeWithPrefix('tsource.', columns);

    const sql = ` when not matched then insert (${destinationColumns}) values (${sourceColumns})`;

    return sql;
  }

  _getParameters(this: any, params: any) {
    const sql = this.client.parameterize(
      params,
      this.client.valueForUndefined,
      this.builder,
      this.bindingsHolder,
    );

    return sql;
  }

  _mergeInsertIsEmpty(this: any, insert: any) {
    return (Array.isArray(insert) && insert.length === 0)
      || (typeof insert === 'object' && Utils.isEmpty(insert));
  }

  _mergeOn(this: any, conflict: any) {
    let sql = 'on 1=1';

    if (Array.isArray(conflict)) {
      const conflictColumn = this.formatter.columnize(conflict[0]);
      sql = `on ${this.tableName}.${conflictColumn} = tsource.${conflictColumn}`;
    }

    return sql;
  }

  _insertWithMerge(this: any) {
    const { insert = [], onConflict, ignore, merge, returning, options = {} } = this.single;
    if (this._mergeInsertIsEmpty(insert)) {
      return '';
    }

    const insertData = this._prepInsert(insert);
    const insertParameters = insertData.values.map((value: any) => `(${this._getParameters(value)})`).join(', ');
    const sourceColumns = this.formatter.columnize(insertData.columns);

    const returningSql = returning
      ? ` ${this._returning('insert', returning, options.includeTriggerModifications)}`
      : '';

    let sql = `merge into ${this.tableName} using (values ${insertParameters}) as tsource(${sourceColumns}) `;

    sql += this._mergeOn(onConflict);

    sql += this._mergeWhenNotMatched(insertData.columns);

    if (!ignore) {
      sql += this._mergeWhenMatched(insertData.columns, merge.updates);
    }

    sql += returningSql;

    if (options.includeTriggerModifications) {
      sql = this._buildTempTable(returning) + sql + this._buildReturningSelect(returning);
    }

    sql = this.with() + sql + ';';

    return sql;
  }

}
