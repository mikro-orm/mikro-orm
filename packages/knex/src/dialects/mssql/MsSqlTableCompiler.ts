import { MonkeyPatchable } from '../../MonkeyPatchable';

export class MsSqlTableCompiler extends MonkeyPatchable.MsSqlTableCompiler {

  lowerCase = true;
  addColumnsPrefix = 'add ';
  dropColumnPrefix = 'drop column ';
  alterColumnPrefix = 'alter column ';

  alterColumns(this: any, columns: any, colBuilder: any) {
    for (let i = 0, l = colBuilder.length; i < l; i++) {
      const builder = colBuilder[i];
      if (builder.modified.defaultTo) {
        const schema = this.schemaNameRaw || 'dbo';
        const baseQuery = `declare @constraint${i} varchar(100) = (select default_constraints.name from sys.all_columns`
          + ' join sys.tables on all_columns.object_id = tables.object_id'
          + ' join sys.schemas on tables.schema_id = schemas.schema_id'
          + ' join sys.default_constraints on all_columns.default_object_id = default_constraints.object_id'
          + ` where schemas.name = '${schema}' and tables.name = '${this.tableNameRaw}' and all_columns.name = '${builder.getColumnName()}')`
          + ` if @constraint${i} is not null exec('alter table ${this.tableNameRaw} drop constraint ' + @constraint${i})`;
        this.pushQuery(baseQuery);
      }
    }
    // in SQL server only one column can be altered at a time
    columns.sql.forEach((sql: string) => {
      this.pushQuery({
        sql: `alter table ${this.tableName()} ${this.alterColumnPrefix.toLowerCase()}${sql}`,
        bindings: columns.bindings,
      });
    });
  }

  dropForeign(this: any, columns: any, constraintName: any) {
    constraintName = constraintName
      ? this.formatter.wrap(constraintName)
      : this._indexCommand('foreign', this.tableNameRaw, columns);
    this.pushQuery(`alter table ${this.tableName()} drop constraint ${constraintName}`);
  }

}
