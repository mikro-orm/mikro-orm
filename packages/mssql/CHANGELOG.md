# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [6.4.1](https://github.com/mikro-orm/mikro-orm/compare/v6.4.0...v6.4.1) (2024-12-08)

**Note:** Version bump only for package @mikro-orm/mssql





# [6.4.0](https://github.com/mikro-orm/mikro-orm/compare/v6.3.13...v6.4.0) (2024-11-11)


### Features

* **mssql:** add support for smalldatetime type ([#6128](https://github.com/mikro-orm/mikro-orm/issues/6128)) ([998509b](https://github.com/mikro-orm/mikro-orm/commit/998509bf22dcd85f87e2506323ca4e3ceee052f6)), closes [#5904](https://github.com/mikro-orm/mikro-orm/issues/5904)





## [6.3.13](https://github.com/mikro-orm/mikro-orm/compare/v6.3.12...v6.3.13) (2024-10-06)

**Note:** Version bump only for package @mikro-orm/mssql





## [6.3.12](https://github.com/mikro-orm/mikro-orm/compare/v6.3.11...v6.3.12) (2024-10-01)

**Note:** Version bump only for package @mikro-orm/mssql





## [6.3.11](https://github.com/mikro-orm/mikro-orm/compare/v6.3.10...v6.3.11) (2024-09-26)


### Bug Fixes

* **core:** support atomic updates on properties with custom types ([951284b](https://github.com/mikro-orm/mikro-orm/commit/951284bd66457a74ae7c2eca5a87f7918a0ecd13)), closes [#6030](https://github.com/mikro-orm/mikro-orm/issues/6030)





## [6.3.10](https://github.com/mikro-orm/mikro-orm/compare/v6.3.9...v6.3.10) (2024-09-15)

**Note:** Version bump only for package @mikro-orm/mssql





## [6.3.9](https://github.com/mikro-orm/mikro-orm/compare/v6.3.8...v6.3.9) (2024-09-09)

**Note:** Version bump only for package @mikro-orm/mssql





## [6.3.8](https://github.com/mikro-orm/mikro-orm/compare/v6.3.7...v6.3.8) (2024-09-04)

**Note:** Version bump only for package @mikro-orm/mssql





## [6.3.7](https://github.com/mikro-orm/mikro-orm/compare/v6.3.6...v6.3.7) (2024-08-28)

**Note:** Version bump only for package @mikro-orm/mssql





## [6.3.6](https://github.com/mikro-orm/mikro-orm/compare/v6.3.5...v6.3.6) (2024-08-14)

**Note:** Version bump only for package @mikro-orm/mssql





## [6.3.5](https://github.com/mikro-orm/mikro-orm/compare/v6.3.4...v6.3.5) (2024-08-11)


### Bug Fixes

* **entity-generator:** ensure `columnType` is emitted correctly and when necessary ([#5930](https://github.com/mikro-orm/mikro-orm/issues/5930)) ([72333ad](https://github.com/mikro-orm/mikro-orm/commit/72333ad506129834185f39d44457fdfb947df35c)), closes [#5928](https://github.com/mikro-orm/mikro-orm/issues/5928)
* **postgres:** respect `timezone` option and interpret `timestamp` columns in UTC by default ([#5916](https://github.com/mikro-orm/mikro-orm/issues/5916)) ([1da0722](https://github.com/mikro-orm/mikro-orm/commit/1da07220ea5ac3f70b9244605b5e6367cf8f8d0c)), closes [#5591](https://github.com/mikro-orm/mikro-orm/issues/5591)





## [6.3.4](https://github.com/mikro-orm/mikro-orm/compare/v6.3.3...v6.3.4) (2024-08-06)


### Bug Fixes

* **mssql:** convert tuple comparison queries to simple `and/or` conditions ([#5906](https://github.com/mikro-orm/mikro-orm/issues/5906)) ([c3c3519](https://github.com/mikro-orm/mikro-orm/commit/c3c3519db72ab15810fcb65d764541ab1fcc0130))





## [6.3.3](https://github.com/mikro-orm/mikro-orm/compare/v6.3.2...v6.3.3) (2024-08-03)

**Note:** Version bump only for package @mikro-orm/mssql





## [6.3.2](https://github.com/mikro-orm/mikro-orm/compare/v6.3.1...v6.3.2) (2024-08-01)

**Note:** Version bump only for package @mikro-orm/mssql





## [6.3.1](https://github.com/mikro-orm/mikro-orm/compare/v6.3.0...v6.3.1) (2024-07-25)


### Bug Fixes

* **mssql:** do not escape charecters that don't need escaping ([#5860](https://github.com/mikro-orm/mikro-orm/issues/5860)) ([6730978](https://github.com/mikro-orm/mikro-orm/commit/6730978b6c7003e45ed97f9d2d7c5c0bc9262e26)), closes [mikro-orm#5811](https://github.com/mikro-orm/issues/5811)
* **mssql:** do not escape new line character ([97919ce](https://github.com/mikro-orm/mikro-orm/commit/97919ce479fc5e2e00a3bf24a511ec29a2279d1b)), closes [#5811](https://github.com/mikro-orm/mikro-orm/issues/5811)





# [6.3.0](https://github.com/mikro-orm/mikro-orm/compare/v6.2.9...v6.3.0) (2024-07-18)


### Bug Fixes

* **core:** improve handling of JSON properties to support numeric strings in all drivers ([#5780](https://github.com/mikro-orm/mikro-orm/issues/5780)) ([fc50c5f](https://github.com/mikro-orm/mikro-orm/commit/fc50c5f5f28f0764115631900edac24bc734afa4)), closes [#5773](https://github.com/mikro-orm/mikro-orm/issues/5773)
* **mssql:** add proper support for MSSQL's native "varchar" type ([#5685](https://github.com/mikro-orm/mikro-orm/issues/5685)) ([0b514ce](https://github.com/mikro-orm/mikro-orm/commit/0b514ce7378df21ef414027f993267f2ecbe681a))
* **mssql:** fix handling of non-UTC timezones ([e78696c](https://github.com/mikro-orm/mikro-orm/commit/e78696c548e14be3e00ba5595697816db4c9dd52)), closes [#5695](https://github.com/mikro-orm/mikro-orm/issues/5695)
* **mssql:** only escape strings and unicode strings when necessary ([#5786](https://github.com/mikro-orm/mikro-orm/issues/5786)) ([b4e0914](https://github.com/mikro-orm/mikro-orm/commit/b4e0914772356285bcbb9362a4df2044438b4cd7)), closes [#5811](https://github.com/mikro-orm/mikro-orm/issues/5811)


### Features

* **core:** addz `Platform.getDefaultVarcharLength` and optional `Type.getDefaultLength` ([#5749](https://github.com/mikro-orm/mikro-orm/issues/5749)) ([29dcdeb](https://github.com/mikro-orm/mikro-orm/commit/29dcdeb5e4c3f84e43c154fe3eb81a113c6d1470))
* **core:** implement "character" type (DB type "char") ([#5684](https://github.com/mikro-orm/mikro-orm/issues/5684)) ([9fa5fad](https://github.com/mikro-orm/mikro-orm/commit/9fa5fad5e3955cdcdee89aa12c8b3dd4841b2045))
* **query-builder:** infer `Loaded` hint based on `joinAndSelect` calls ([#5482](https://github.com/mikro-orm/mikro-orm/issues/5482)) ([d18da6b](https://github.com/mikro-orm/mikro-orm/commit/d18da6b8cfce84ffaf480a27b869b79efbc70fb6))





## [6.2.9](https://github.com/mikro-orm/mikro-orm/compare/v6.2.8...v6.2.9) (2024-05-31)


### Bug Fixes

* **mssql:** account for quotes in table names ([#5637](https://github.com/mikro-orm/mikro-orm/issues/5637)) ([0343609](https://github.com/mikro-orm/mikro-orm/commit/0343609c3e99d1fe5e4024a7afe0aaa2b2d6d980))
* **mssql:** fix creating migrations due to a missing helper method ([#5644](https://github.com/mikro-orm/mikro-orm/issues/5644)) ([90e27c2](https://github.com/mikro-orm/mikro-orm/commit/90e27c275d7db46269ee721e049dcfe09b274abe)), closes [#5633](https://github.com/mikro-orm/mikro-orm/issues/5633)
* **mssql:** fix ensuring the database exists on older SQL Server versions ([f0a5790](https://github.com/mikro-orm/mikro-orm/commit/f0a5790de0b08978983a3af82122e0f5045531dc)), closes [#5638](https://github.com/mikro-orm/mikro-orm/issues/5638)





## [6.2.8](https://github.com/mikro-orm/mikro-orm/compare/v6.2.7...v6.2.8) (2024-05-21)

**Note:** Version bump only for package @mikro-orm/mssql





## [6.2.7](https://github.com/mikro-orm/mikro-orm/compare/v6.2.6...v6.2.7) (2024-05-18)

**Note:** Version bump only for package @mikro-orm/mssql





## [6.2.6](https://github.com/mikro-orm/mikro-orm/compare/v6.2.5...v6.2.6) (2024-05-14)

**Note:** Version bump only for package @mikro-orm/mssql





## [6.2.5](https://github.com/mikro-orm/mikro-orm/compare/v6.2.4...v6.2.5) (2024-05-05)

**Note:** Version bump only for package @mikro-orm/mssql





## [6.2.4](https://github.com/mikro-orm/mikro-orm/compare/v6.2.3...v6.2.4) (2024-05-02)

**Note:** Version bump only for package @mikro-orm/mssql





## [6.2.3](https://github.com/mikro-orm/mikro-orm/compare/v6.2.2...v6.2.3) (2024-04-24)

**Note:** Version bump only for package @mikro-orm/mssql





## [6.2.2](https://github.com/mikro-orm/mikro-orm/compare/v6.2.1...v6.2.2) (2024-04-20)


### Bug Fixes

* **mssql:** declare `import` types explicitly ([02494bf](https://github.com/mikro-orm/mikro-orm/commit/02494bf129e15261ba17218288a840120fedb089)), closes [#5462](https://github.com/mikro-orm/mikro-orm/issues/5462)





## [6.2.1](https://github.com/mikro-orm/mikro-orm/compare/v6.2.0...v6.2.1) (2024-04-12)


### Bug Fixes

* **mssql:** support instance names in `host` ([dc7dc4c](https://github.com/mikro-orm/mikro-orm/commit/dc7dc4cecfc8e3c426ce8c679365c8efa7705370)), closes [#5441](https://github.com/mikro-orm/mikro-orm/issues/5441)





# [6.2.0](https://github.com/mikro-orm/mikro-orm/compare/v6.1.12...v6.2.0) (2024-04-09)


### Features

* **mssql:** add MS SQL Server driver ([#1375](https://github.com/mikro-orm/mikro-orm/issues/1375)) ([eeaad45](https://github.com/mikro-orm/mikro-orm/commit/eeaad45a60b3ef4732d5ba9eafc8719998e52181)), closes [#771](https://github.com/mikro-orm/mikro-orm/issues/771)
