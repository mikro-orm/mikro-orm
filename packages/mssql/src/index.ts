/* istanbul ignore file */
export * from '@mikro-orm/knex';
export * from './MsSqlConnection';
export * from './MsSqlDriver';
export * from './MsSqlPlatform';
export * from './MsSqlSchemaHelper';
export * from './MsSqlExceptionConverter';
export * from './UnicodeStringType';
export {
  MsSqlMikroORM as MikroORM,
  MsSqlOptions as Options,
  defineMsSqlConfig as defineConfig,
} from './MsSqlMikroORM';
