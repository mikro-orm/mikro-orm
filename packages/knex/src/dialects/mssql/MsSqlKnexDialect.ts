import { MonkeyPatchable } from '../../MonkeyPatchable';
import { MsSqlTableCompiler } from './MsSqlTableCompiler';
import { MsSqlColumnCompiler } from './MsSqlColumnCompiler';
import { MsSqlQueryCompiler } from './MsSqlQueryCompiler';

export class MsSqlKnexDialect extends MonkeyPatchable.MsSqlDialect {

  tableCompiler() {
    // eslint-disable-next-line prefer-rest-params
    return new (MsSqlTableCompiler as any)(this, ...arguments);
  }

  columnCompiler() {
    // eslint-disable-next-line prefer-rest-params
    return new (MsSqlColumnCompiler as any)(this, ...arguments);
  }

  queryCompiler() {
    // eslint-disable-next-line prefer-rest-params
    return new (MsSqlQueryCompiler as any)(this, ...arguments);
  }

}
