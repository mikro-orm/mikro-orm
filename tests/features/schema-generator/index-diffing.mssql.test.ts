import { Entity, Ref, Index, ManyToOne, MikroORM, PrimaryKey, Property, Unique } from '@mikro-orm/mssql';

@Entity()
class Author {

  @PrimaryKey()
  id!: number;

  @Property()
  name!: string;

}

@Entity({ tableName: 'book' })
class Book1 {

  @PrimaryKey()
  id!: number;

  @ManyToOne(() => Author, { ref: true })
  author1!: Ref<Author>;

  @ManyToOne(() => Author, { ref: true })
  author2!: Ref<Author>;

  @ManyToOne(() => Author)
  author3!: Author;

  @ManyToOne(() => Author)
  author4!: Author;

  @ManyToOne(() => Author)
  author5!: Author;

  @Property()
  title!: string;

  @Property({ unique: true })
  isbn!: string;

}

@Entity({ tableName: 'book' })
@Index({ properties: 'author1' })
@Index({ properties: 'author3' })
@Index({ name: 'custom_index_expr123', expression: 'CREATE  INDEX  [custom_index_expr123] ON [book] ([isbn]) WHERE [isbn] IS NOT NULL' })
class Book2 {

  @PrimaryKey()
  id!: number;

  @ManyToOne(() => Author, { ref: true })
  author1!: Ref<Author>;

  @ManyToOne(() => Author, { ref: true })
  @Index()
  author2!: Ref<Author>;

  @ManyToOne(() => Author)
  author3!: Author;

  @ManyToOne(() => Author)
  @Index()
  author4!: Author;

  @ManyToOne(() => Author, { index: true })
  author5!: Author;

  @Index({ expression: 'CREATE INDEX [custom_index_expr] ON [book] ([title])' })
  @Property()
  title!: string;

  @Property({ unique: 'isbn_unique_constr' })
  isbn!: string;

}

@Entity({ tableName: 'book' })
@Index({ properties: 'author1' })
@Index({ properties: 'author3', name: 'lol31' })
@Index({ properties: 'author3', name: 'lol41' })
@Index({ name: 'custom_index_expr123', expression: 'create  index  [custom_index_expr123] on [book] ([isbn]) WHERE [isbn] IS NOT NULL' })
class Book3 {

  @PrimaryKey()
  id!: number;

  @ManyToOne(() => Author, { ref: true })
  author1!: Ref<Author>;

  @ManyToOne(() => Author, { ref: true })
  @Index()
  author2!: Ref<Author>;

  @ManyToOne(() => Author)
  author3!: Author;

  @ManyToOne(() => Author)
  @Index()
  author4!: Author;

  @ManyToOne(() => Author, { index: 'auth_idx5' })
  author5!: Author;

  @Index({ name: 'custom_index_expr2', expression: 'create index [custom_index_expr2] on [book] ([title])' })
  @Property()
  title!: string;

  @Property()
  @Unique()
  isbn!: string;

}

@Entity({ tableName: 'book' })
@Index({ properties: 'author1' })
@Index({ properties: 'author3', name: 'lol32' })
@Index({ properties: 'author3', name: 'lol42' })
class Book4 {

  @PrimaryKey()
  id!: number;

  @ManyToOne(() => Author, { ref: true })
  author1!: Ref<Author>;

  @ManyToOne(() => Author, { ref: true })
  @Index()
  author2!: Ref<Author>;

  @ManyToOne(() => Author)
  author3!: Author;

  @ManyToOne(() => Author)
  @Index()
  author4!: Author;

  @ManyToOne(() => Author, { index: 'auth_idx5' })
  author5!: Author;

  @Property()
  title!: string;

  @Property()
  @Unique()
  isbn!: string;

}

describe('indexes on FKs in postgres (GH 1518)', () => {

  let orm: MikroORM;

  beforeAll(async () => {
    orm = await MikroORM.init({
      entities: [Author],
      dbName: `mikro_orm_test_gh_1518`,
      password: 'Root.Root',
    });

    await orm.schema.refreshDatabase({ dropDb: true });
  });

  afterAll(() => orm.close(true));

  test('schema generator respect indexes on FKs on column update', async () => {
    await orm.discoverEntity(Book1);
    orm.getMetadata().reset('Book0');
    const diff1 = await orm.schema.getUpdateSchemaSQL({ wrap: false });
    expect(diff1).toMatchSnapshot();
    await orm.schema.execute(diff1);

    orm.getMetadata().reset('Book1');
    await orm.discoverEntity(Book2);
    const diff2 = await orm.schema.getUpdateSchemaSQL({ wrap: false });
    expect(diff2).toMatchSnapshot();
    await orm.schema.execute(diff2);

    orm.getMetadata().reset('Book2');
    await orm.discoverEntity(Book3);
    const diff3 = await orm.schema.getUpdateSchemaSQL({ wrap: false });
    expect(diff3).toMatchSnapshot();
    await orm.schema.execute(diff3);

    orm.getMetadata().reset('Book3');
    await orm.discoverEntity(Book4);
    const diff4 = await orm.schema.getUpdateSchemaSQL({ wrap: false });
    expect(diff4).toMatchSnapshot();
    await orm.schema.execute(diff4);
  });

});
