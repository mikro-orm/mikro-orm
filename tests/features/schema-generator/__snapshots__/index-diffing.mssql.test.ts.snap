// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`indexes on FKs in postgres (GH 1518) schema generator respect indexes on FKs on column update 1`] = `
"CREATE TABLE [book] ([id] int identity(1,1) not null primary key, [author1_id] int not null, [author2_id] int not null, [author3_id] int not null, [author4_id] int not null, [author5_id] int not null, [title] nvarchar(255) not null, [isbn] nvarchar(255) not null);
CREATE UNIQUE INDEX [book_isbn_unique] ON [book] ([isbn]) WHERE [isbn] IS NOT NULL;

alter table [book] add constraint [book_author1_id_foreign] foreign key ([author1_id]) references [author] ([id]) on update cascade;
alter table [book] add constraint [book_author2_id_foreign] foreign key ([author2_id]) references [author] ([id]);
alter table [book] add constraint [book_author3_id_foreign] foreign key ([author3_id]) references [author] ([id]);
alter table [book] add constraint [book_author4_id_foreign] foreign key ([author4_id]) references [author] ([id]);
alter table [book] add constraint [book_author5_id_foreign] foreign key ([author5_id]) references [author] ([id]);

"
`;

exports[`indexes on FKs in postgres (GH 1518) schema generator respect indexes on FKs on column update 2`] = `
"DROP INDEX [book_isbn_unique] ON [book];

CREATE INDEX [book_author1_id_index] ON [book] ([author1_id]);
CREATE INDEX [book_author2_id_index] ON [book] ([author2_id]);
CREATE INDEX [book_author3_id_index] ON [book] ([author3_id]);
CREATE INDEX [book_author4_id_index] ON [book] ([author4_id]);
CREATE INDEX [book_author5_id_index] ON [book] ([author5_id]);
CREATE UNIQUE INDEX [isbn_unique_constr] ON [book] ([isbn]) WHERE [isbn] IS NOT NULL;
CREATE INDEX [custom_index_expr] ON [book] ([title]);
CREATE  INDEX  [custom_index_expr123] ON [book] ([isbn]) WHERE [isbn] IS NOT NULL;

"
`;

exports[`indexes on FKs in postgres (GH 1518) schema generator respect indexes on FKs on column update 3`] = `
"DROP INDEX [custom_index_expr] ON [book];
DROP INDEX [isbn_unique_constr] ON [book];

CREATE UNIQUE INDEX [book_isbn_unique] ON [book] ([isbn]) WHERE [isbn] IS NOT NULL;
create index [custom_index_expr2] on [book] ([title]);
CREATE INDEX [lol31] ON [book] ([author3_id]);
drop index [book_author3_id_index] on [dbo].[book];
create index [lol41] on [dbo].[book] ([author3_id]);
drop index [book_author5_id_index] on [dbo].[book];
create index [auth_idx5] on [dbo].[book] ([author5_id]);

"
`;

exports[`indexes on FKs in postgres (GH 1518) schema generator respect indexes on FKs on column update 4`] = `
"DROP INDEX [custom_index_expr123] ON [book];
DROP INDEX [custom_index_expr2] ON [book];
DROP INDEX [lol31] ON [book];
DROP INDEX [lol41] ON [book];

CREATE INDEX [lol42] ON [book] ([author3_id]);
CREATE INDEX [lol32] ON [book] ([author3_id]);

"
`;
